### ISSUE TITLE
Problem issuing certificates - 200 self test failed - Nginx

### ISSUE THREAD

**Issue by tomwaldnz** [hlandau/acme#249]
I'd appreciate some help diagnosing why certificates aren't being issued. I seem to run into problems like this occasionally, so I'd really like to understand this and get it sorted once and for all.

The core problem is I'm getting "self test failed", but I can't see why - the logs don't have any information that I can see.

I run 5 domains on my server, but today I'm only concerned with wildphotography.co.nz and the www subdomain. Acmetool is pointing at the LE staging server, and is set up for webroot validation - I understand that LE tries every method it can though.

I set the Let's Encrypt webroot to /var/www/.well-known/acme-challenge/

Here's what I believe is the relevant part of my Nginx web server configuration

```
server {
    listen       80;
    server_name  www.wildphotography.co.nz wildphotography.co.nz;
    
    location /.well-known/acme-challenge/ {
      alias /var/www/.well-known/acme-challenge/;
    }

    location / {
      return       301 https://www.wildphotography.co.nz$request_uri;
    }
}
```
This shows the web server is serving files over http. I created the file /var/www/.well-known/acme-challenge/hello.html with the contents "hello, world", which I can access with a web browser from my remote PC and using a curl on the host

```
curl -i  http://www.wildphotography.co.nz/.well-known/acme-challenge/hello.html
HTTP/1.1 200 OK
Server: nginx
Date: Wed, 05 Apr 2017 20:37:29 GMT
Content-Type: text/html
Content-Length: 13
Last-Modified: Wed, 05 Apr 2017 20:20:26 GMT
Connection: keep-alive
ETag: "58e5518a-d"
Accept-Ranges: bytes

hello, world
```

Here's the corresponding Nginx access log entry

`11.0.1.10 [06/Apr/2017:08:50:48 +1200] "GET /.well-known/acme-challenge/hello.html HTTP/1.1" 200 13 "-" "curl/7.51.0" www.wildphotography.co.nz 0.000 - http`

I set up folder monitoring to check if acmetool has created the challenge files

`inotifywait -m -r /var/www/.well-known/acme-challenge`

I then run acmetool

`/usr/local/bin/acmetool --xlog.severity=debug > /tmp/dump14 2>&1 want www.wildphotography.co.nz wildphotography.co.nz`

inotifywait shows that the challenge files have been written, but not read. I'm including the hello.html entry to show that it is working

```
./ OPEN hello.html
./ ACCESS hello.html
./ CREATE pqbw55n9r8zcD609peSj7xlD1AOzT95ZEf1yi5jB1gM
./ OPEN pqbw55n9r8zcD609peSj7xlD1AOzT95ZEf1yi5jB1gM
./ MODIFY pqbw55n9r8zcD609peSj7xlD1AOzT95ZEf1yi5jB1gM
./ CLOSE_WRITE,CLOSE pqbw55n9r8zcD609peSj7xlD1AOzT95ZEf1yi5jB1gM
./ DELETE pqbw55n9r8zcD609peSj7xlD1AOzT95ZEf1yi5jB1gM
```

And then I parse the output, as someone told me on a previous question a while back

fgrep -v fdb: /tmp/dump14 | fgrep -v storageops: > dump15

Here are the lines from the parsed log

```
20170406084635 [DEBUG] acme.solver: attempting challenge type http-01
20170406084635 [DEBUG] acme.responder: failed to listen on [::]:80: listen tcp [::]:80: bind: address already in use
20170406084635 [DEBUG] acme.responder: failed to listen on :80: listen tcp :80: bind: address already in use
20170406084635 [DEBUG] acme.responder: listening on [::1]:402
20170406084635 [DEBUG] acme.responder: listening on 127.0.0.1:402
20170406084635 [DEBUG] acme.responder: listening on [::1]:4402
20170406084635 [DEBUG] acme.responder: listening on 127.0.0.1:4402
20170406084635 [DEBUG] acme.responder: writing 2 webroot challenge files
20170406084635 [DEBUG] acme.responder: writing webroot file /var/www/.well-known/acme-challenge/pqbw55n9r8zcD609peSj7xlD1AOzT95ZEf1yi5jB1gM
20170406084635 [DEBUG] acme.responder: writing webroot file /var/run/acme/acme-challenge/pqbw55n9r8zcD609peSj7xlD1AOzT95ZEf1yi5jB1gM
20170406084635 [DEBUG] acme.hooks: calling hook script: /usr/libexec/acme/hooks/haproxy
20170406084635 [DEBUG] acme.hooks: calling hook script: /usr/libexec/acme/hooks/reload
20170406084635 [DEBUG] acme.responder: http-01 self test
20170406084635 [INFO] acme.responder: http-01 self test failed: non-200 status code when doing self-test
20170406084635 [DEBUG] acme.responder: removing webroot file /var/www/.well-known/acme-challenge/pqbw55n9r8zcD609peSj7xlD1AOzT95ZEf1yi5jB1gM
20170406084635 [DEBUG] acme.responder: removing webroot file /var/run/acme/acme-challenge/pqbw55n9r8zcD609peSj7xlD1AOzT95ZEf1yi5jB1gM
20170406084635 [DEBUG] acme.hooks: calling hook script: /usr/libexec/acme/hooks/haproxy
20170406084635 [DEBUG] acme.hooks: calling hook script: /usr/libexec/acme/hooks/reload
20170406084635 [DEBUG] acme.solver: challenge start failed: non-200 status code when doing self-test
20170406084635 [DEBUG] acme.solver: attempting challenge type dns-01
20170406084635 [DEBUG] acme.hooks: calling hook script: /usr/libexec/acme/hooks/haproxy
20170406084635 [DEBUG] acme.hooks: calling hook script: /usr/libexec/acme/hooks/reload

```

There is no request made to Nginx for that file

```
vi /var/log/nginx/wp.access.log
/pqbw55
E486: Pattern not found: pqbw55
```

Here's the acmetool --version output, which shows v0.0.59.

```
/usr/local/bin/acmetool --version
go version go1.7 linux/386 gc cgo=false
built by travis
git github.com/alecthomas/template a0175ee3bccc567396460bf5acd36800cb10c49c heads/master
git github.com/alecthomas/units 2efee857e7cfd4f3d0138cc3cbb1b4966962b93a heads/master
git github.com/coreos/go-systemd e97b35f834b17eaa82afe3d44715c34736bfa12b heads/master
git github.com/godbus/dbus 4b24ebee04561bf8a3bcc09aead82062edc56778 heads/master
git github.com/hlandau/acme 3201252d3de450a556a33daedbb55b95689eb248 v0.0.59
git github.com/hlandau/buildinfo 337a29b5499734e584d4630ce535af64c5fe7813 heads/master
git github.com/hlandau/dexlogconfig 244f29bd260884993b176cd14ef2f7631f6f3c18 heads/master
git github.com/hlandau/goutils 0cdb66aea5b843822af6fdffc21286b8fe8379c4 heads/master
git github.com/hlandau/xlog 197ef798aed28e08ed3e176e678fda81be993a31 v1.0.0
git github.com/jmhodges/clock 880ee4c335489bc78d01e4d0a254ae880734bc15 tags/v1.1
git github.com/mattn/go-isatty dda3de49cbfcec471bd7a70e6cc01fcc3ff90109 heads/master
git github.com/mattn/go-runewidth 14207d285c6c197daabb5c9793d63e7af9ab2d50 heads/master
git github.com/mitchellh/go-wordwrap ad45545899c7b13c020ea92b2072220eefad42b8 heads/master
git github.com/ogier/pflag 45c278ab3607870051a2ea9040bb85fcb8557481 heads/master
git github.com/peterhellberg/link 24c1495e8c97c8c537f23307b2b8d2932051c1a9 heads/master
git github.com/satori/go.uuid b061729afc07e77a8aa4fad0a2fd840958f1942a heads/master
git github.com/shiena/ansicolor a422bbe96644373c5753384a59d678f7d261ff10 heads/master
git golang.org/x/crypto 453249f01cfeb54c3d549ddb75ff152ca243f9d8 heads/master
git golang.org/x/net b4690f45fa1cafc47b1c280c2e75116efe40cc13 heads/master
git gopkg.in/alecthomas/kingpin.v2 e9044be3ab2a8e11d4e1f418d12f0790d57e8d70 tags/v2.2.3
git gopkg.in/cheggaaa/pb.v1 d7e6ca3010b6f084d8056847f55d7f572f180678 v1.0.7
git gopkg.in/hlandau/configurable.v1 41496864a1fe3e0fef2973f22372b755d2897402 v1.0.1
git gopkg.in/hlandau/easyconfig.v1 33e53e2d08656ccad000531debbf2656a896b695 v1.0.15
git gopkg.in/hlandau/service.v2 b64b3467ebd16f64faec1640c25e318efc0c0d7b v2.0.16
git gopkg.in/hlandau/svcutils.v1 c25dac49e50cbbcbef8c81b089f56156f4067729 v1.0.10
git gopkg.in/square/go-jose.v1 aa2e30fdd1fe9dd3394119af66451ae790d50e0d v1.1.0
git gopkg.in/tylerb/graceful.v1 4654dfbb6ad53cb5e27f37d99b02e16c1872fbbb tags/v1.2.15
git gopkg.in/yaml.v2 a3f3340b5840cee44f372bddb5880fcbc419b46a heads/v2
```

Can anyone help work out why this isn't working?

**Comment by hlandau** [comment]
Might want to test curl for the non-www domain as well.

Nothing jumps out at me, but you could always try the proxy method instead. You'd need to proxy requests from /.well-known/acme-challenge/ to the same path at 127.0.0.1:402.

Failing that, you can disable HTTP self-test if you like. This will mean any errors provisioning the challenge file will result in validation errors at Let's Encrypt instead, so it's unlikely to be helpful, unless there's some problem with how acmetool is performing the self-test. To do this, edit your targetfile(s) to include the following:
```
request:
  challenge:
    http-self-test: false
```

**Comment by tomwaldnz** [comment]
Thanks for the fast response and the suggestions :) The non-www domain curl does the same thing, 200 code.

Is there any way to determine why the self test fails? Could more logging be added? That it doesn't even try to request the verification file from the web server could be a hint about where it's giving up in the code.

I've just worked out what a targetfile is. I'll read the documentation fully in the next day or so.

One question about targetfiles. I have two files for the wildphotography domain, probably because of experimenting. One contains the root domain and two subdomains, the other contains the root domain and one subdomain. Should I delete the file that contains only two? Is there any risk doing that?

I'll try disabling self test and report back when I've tried it.

I'd rather get this method working, but I'll try the proxy method as well. Maybe it'll be more reliable.

**Comment by tomwaldnz** [comment]
I've gotten this working, but I'm not 100% why as I stupidly made two changes at once. Now it's working I don't want to mess with it too much.

First up I tried the proxy method. It didn't work. It said the same thing as the other method in the logs.

Next I disabled the self test. I also added the /.well-known/acme-challenge/ block to my other domain, proofs.wildphotography.co.nz. I then ran against staging and it worked. So changed to production, ran it, and it worked as well.

I have another domain due for renewal in the next day or so. I'll make sure I have Nginx set up correctly, but I won't disable the self check. If it fails I'll enable the self check.

I should be able to report back in a day or two with a conclusion.

**Comment by tomwaldnz** [comment]
This is curious. I created a new subdomain specifically to test with a fresh domain. I set Nginx up as above, requested a certificate against staging, and it came through. Great.

I changed to production and tried again, didn't get any errors. However it didn't work - when I used curl it gave me certificate errors. A bit of digging showed the certificate chain still appeared to be on the staging server

issuer=/CN=Fake LE Intermediate X1

I deleted the folder containing the certificates and keys, ran it again, and it issued properly against production.

Looks like if you use staging you need to manually delete everything before moving back to production. I wonder if acmetool could be modified so production resources always overwrite testing resources.

**Comment by hlandau** [comment]
It's correct that acmetool doesn't validate certificate chains. It doesn't have any particular understanding of CAs; it just gets certificates from the ACME server you specify. An existing certificate is considered usable so long as the keys for it are on hand and it isn't expired. This is by design; fixing this would require acmetool to understand what ACME servers are "supposed" to issue real certificates, which are "supposed" to issue fake ones, etc.

If you have multiple targetfiles, you should probably just merge them into one if you have a small number of domains (the limit for names on one certificate is 100 with Let's Encrypt, and Let's Encrypt applies rate limits by certificate, not domain, so it's in your interest to minimise the number of certificates issued.)

**Comment by tomwaldnz** [comment]
I have 11 domains if you include subdomains, with the certificates expiring at different times. If I merged them into one I guess it will renew when the first domain needs to be renewed? I'm not coming close to any limits, so I'll probably just leave them alone.

I'll leave this open a bit longer, until my next certificate renews. I want to dig into if I need to disable self checks for existing certificates, or if I had something wrong with my configurations. It should try to renew automatically tomorrow.

As I dig into this more, look in the configurations, read the documentation, this is becoming a lot less opaque. It's still somewhat complex though.

**Comment by hlandau** [comment]
If you merge them into one, it will issue a new certificate immediately, because having multiple names in the same targetfile is considered a demand of acmetool that those names appear on the same certificate. Since no existing certificate you have satisfies this criteria, it will get a new one immediately, so there will only be one certificate acmetool needs to renew in the future.

**Comment by tomwaldnz** [comment]
I would prefer not to have all the URLs on one certificate, as it creates a public association between the sites, if anyone looks. Not that many people would look, or that it's all that important. Interesting to know though.

Today Acmetool automatically renewed the next certificate scheduled to expire. I didn't have to disable self checks. I had changed the directory that was used for the self checks and I modified my Nginx configuration as well. Basically everything in my first post worked.

I don't know why I had the problem, but it seems to be solved now. Thanks Hugo for help, much appreciated. I learned something :)

I wrote a [tutorial on setting up Acmetool for Let's Encrypt on Amazon Linux](https://www.photographerstechsupport.com/tutorials/hosting-wordpress-on-aws-tutorial-part-5-free-https-https2-for-wordpress-using-lets-encrypt-aws/#acme). I've bought it largely up to date, and I'll fix up the downloadable configuration files in the next couple of days.

**Comment by tomwaldnz** [comment]
Another thing I've found, which someone might find useful.

To work out if you've issued against staging or live look at the url file in the correct folder.

`more /var/lib/acme/live/example.com url`

If you issued a certificate using the staging server, the best way I've found to issue against production is to delete all references to the current site. For me that means deleting (all under /var/lib/acmetool)

TAKE A BACKUP FIRST
```
rm -rf /desired/example.com*
rm -rf /live/example.com and /live/www.example.com
rm -rf /certs/ID-GOES-HERE
```
